// datastore-core/esm/src/namespace.js
import { Key } from 'interface-datastore';
import { KeyTransformDatastore } from './keytransform.js';
export class NamespaceDatastore extends KeyTransformDatastore {
  constructor(child, prefix) {
    super(child, {
      convert(key) {
        return prefix.child(key);
      },
      invert(key) {
        if (prefix.toString() === '/') {
          return key;
        }
        if (!prefix.isAncestorOf(key)) {
          throw new Error(`Expected prefix: (${ prefix.toString() }) in key: ${ key.toString() }`);
        }
        return new Key(key.toString().slice(prefix.toString().length), false);
      }
    });
    this.prefix = prefix;
  }
  query(q, options) {
    if (q.prefix && this.prefix.toString() !== '/') {
      return super.query(Object.assign({}, q, { prefix: this.prefix.child(new Key(q.prefix)).toString() }));
    }
    return super.query(q, options);
  }
  queryKeys(q, options) {
    if (q.prefix && this.prefix.toString() !== '/') {
      return super.queryKeys(Object.assign({}, q, { prefix: this.prefix.child(new Key(q.prefix)).toString() }));
    }
    return super.queryKeys(q, options);
  }
}
