// ipfs-core/esm/src/components/files/utils/to-mfs-path.js
import { loadMfsRoot } from './with-mfs-root.js';
import { toPathComponents } from './to-path-components.js';
import { exporter } from 'ipfs-unixfs-exporter';
import errCode from 'err-code';
import { CID } from 'multiformats/cid';
const IPFS_PREFIX = 'ipfs';
export const toMfsPath = async (context, path, options) => {
  const root = await loadMfsRoot(context, options);
  let output = { entryType: 'file' };
  let ipfsPath = '';
  if (CID.asCID(path)) {
    ipfsPath = `/ipfs/${ path }`;
  } else {
    ipfsPath = path.toString();
  }
  ipfsPath = ipfsPath.trim();
  ipfsPath = ipfsPath.replace(/(\/\/+)/g, '/');
  if (ipfsPath.endsWith('/') && ipfsPath.length > 1) {
    ipfsPath = ipfsPath.substring(0, ipfsPath.length - 1);
  }
  if (!ipfsPath) {
    throw errCode(new Error('paths must not be empty'), 'ERR_NO_PATH');
  }
  if (ipfsPath.substring(0, 1) !== '/') {
    throw errCode(new Error('paths must start with a leading slash'), 'ERR_INVALID_PATH');
  }
  if (ipfsPath.substring(ipfsPath.length - 1) === '/') {
    ipfsPath = ipfsPath.substring(0, ipfsPath.length - 1);
  }
  const pathComponents = toPathComponents(ipfsPath);
  if (pathComponents[0] === IPFS_PREFIX) {
    let mfsDirectory;
    if (pathComponents.length === 2) {
      mfsDirectory = `/${ pathComponents.join('/') }`;
    } else {
      mfsDirectory = `/${ pathComponents.slice(0, pathComponents.length - 1).join('/') }`;
    }
    output = {
      type: 'ipfs',
      depth: pathComponents.length - 2,
      entryType: 'file',
      mfsPath: `/${ pathComponents.join('/') }`,
      mfsDirectory,
      parts: pathComponents,
      path: `/${ pathComponents.join('/') }`,
      name: pathComponents[pathComponents.length - 1]
    };
  } else {
    const mfsPath = `/${ IPFS_PREFIX }/${ root }${ pathComponents.length ? '/' + pathComponents.join('/') : '' }`;
    const mfsDirectory = `/${ IPFS_PREFIX }/${ root }/${ pathComponents.slice(0, pathComponents.length - 1).join('/') }`;
    output = {
      type: 'mfs',
      depth: pathComponents.length,
      entryType: 'file',
      mfsDirectory,
      mfsPath,
      parts: pathComponents,
      path: `/${ pathComponents.join('/') }`,
      name: pathComponents[pathComponents.length - 1]
    };
  }
  const cidPath = output.type === 'mfs' ? output.mfsPath : output.path;
  try {
    const res = await exporter(cidPath, context.repo.blocks, options);
    output.cid = res.cid;
    output.mfsPath = `/ipfs/${ res.path }`;
    output.entryType = res.type;
    output.content = res.content;
    if ((output.entryType === 'file' || output.entryType === 'directory') && (res.type === 'file' || res.type === 'directory')) {
      output.unixfs = res.unixfs;
    }
  } catch (err) {
    if (err.code !== 'ERR_NOT_FOUND') {
      throw err;
    }
  }
  output.exists = Boolean(output.cid);
  return output;
};
