// ipfs-core/esm/src/components/files/utils/update-tree.js
import debug from 'debug';
import { addLink } from './add-link.js';
import { decode } from '@ipld/dag-pb';
const log = debug('ipfs:mfs:utils:update-tree');
const defaultOptions = { shardSplitThreshold: 1000 };
export async function updateTree(context, trail, options) {
  options = Object.assign({}, defaultOptions, options);
  log('Trail', trail);
  trail = trail.slice().reverse();
  let index = 0;
  let child;
  for await (const block of context.repo.blocks.getMany(trail.map(node => node.cid))) {
    const node = decode(block);
    const cid = trail[index].cid;
    const name = trail[index].name;
    index++;
    if (!child) {
      child = {
        cid,
        name,
        size: block.length
      };
      continue;
    }
    const result = await addLink(context, {
      parent: node,
      name: child.name,
      cid: child.cid,
      size: child.size,
      flush: options.flush,
      shardSplitThreshold: options.shardSplitThreshold,
      hashAlg: options.hashAlg,
      cidVersion: options.cidVersion
    });
    child = {
      cid: result.cid,
      name,
      size: result.size
    };
  }
  const {cid} = child;
  log(`Final CID ${ cid }`);
  return cid;
}
