// ipfs-core-config/cjs/src/repo.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var os = require('os');
var ipfsRepo = require('ipfs-repo');
var path = require('path');
var datastoreFs = require('datastore-fs');
var datastoreLevel = require('datastore-level');
var blockstoreDatastoreAdapter = require('blockstore-datastore-adapter');
var sharding = require('datastore-core/sharding');
var shard = require('datastore-core/shard');
var fs = require('ipfs-repo/locks/fs');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var os__default = /*#__PURE__*/_interopDefaultLegacy(os);
var path__default = /*#__PURE__*/_interopDefaultLegacy(path);

function createRepo(print, codecs, options = {}) {
  const repoPath = options.path || path__default["default"].join(os__default["default"].homedir(), '.jsipfs');
  let lastMigration;
  const onMigrationProgress = options.onMigrationProgress || function (version, percentComplete, message) {
    if (version !== lastMigration) {
      lastMigration = version;
      print(`Migrating repo from v${ version - 1 } to v${ version }`);
    }
    print(`${ percentComplete.toString().padStart(6, ' ') }% ${ message }`);
  };
  return ipfsRepo.createRepo(repoPath, codeOrName => codecs.getCodec(codeOrName), {
    root: new datastoreFs.FsDatastore(repoPath, { extension: '' }),
    blocks: new blockstoreDatastoreAdapter.BlockstoreDatastoreAdapter(new sharding.ShardingDatastore(new datastoreFs.FsDatastore(`${ repoPath }/blocks`, { extension: '.data' }), new shard.NextToLast(2))),
    datastore: new datastoreLevel.LevelDatastore(`${ repoPath }/datastore`),
    keys: new datastoreFs.FsDatastore(`${ repoPath }/keys`),
    pins: new datastoreLevel.LevelDatastore(`${ repoPath }/pins`)
  }, {
    autoMigrate: options.autoMigrate != null ? options.autoMigrate : true,
    onMigrationProgress: onMigrationProgress,
    repoLock: fs.FSLock
  });
}

exports.createRepo = createRepo;
