// ipfs-core/cjs/src/components/config/profiles.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var set = require('just-safe-set');
var getDefaultConfig = require('ipfs-core-config/config');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var set__default = /*#__PURE__*/_interopDefaultLegacy(set);
var getDefaultConfig__default = /*#__PURE__*/_interopDefaultLegacy(getDefaultConfig);

const profiles = {
  server: {
    description: 'Recommended for nodes with public IPv4 address (servers, VPSes, etc.), disables host and content discovery and UPnP in local networks.',
    transform: config => {
      set__default["default"](config, 'Discovery.MDNS.Enabled', false);
      set__default["default"](config, 'Discovery.webRTCStar.Enabled', false);
      config.Swarm = {
        ...config.Swarm || {},
        DisableNatPortMap: true
      };
      return config;
    }
  },
  'local-discovery': {
    description: 'Sets default values to fields affected by `server` profile, enables discovery and UPnP in local networks.',
    transform: config => {
      set__default["default"](config, 'Discovery.MDNS.Enabled', true);
      set__default["default"](config, 'Discovery.webRTCStar.Enabled', true);
      set__default["default"](config, 'Swarm', {
        ...config.Swarm || {},
        DisableNatPortMap: false
      });
      return config;
    }
  },
  test: {
    description: 'Reduces external interference, useful for running ipfs in test environments. Note that with these settings node won\'t be able to talk to the rest of the network without manual bootstrap.',
    transform: config => {
      const defaultConfig = getDefaultConfig__default["default"]();
      set__default["default"](config, 'Addresses.API', defaultConfig.Addresses.API ? '/ip4/127.0.0.1/tcp/0' : '');
      set__default["default"](config, 'Addresses.Gateway', defaultConfig.Addresses.Gateway ? '/ip4/127.0.0.1/tcp/0' : '');
      set__default["default"](config, 'Addresses.Swarm', defaultConfig.Addresses.Swarm.length ? ['/ip4/127.0.0.1/tcp/0'] : []);
      set__default["default"](config, 'Addresses.Delegates', []);
      set__default["default"](config, 'Bootstrap', []);
      set__default["default"](config, 'Discovery.MDNS.Enabled', false);
      set__default["default"](config, 'Discovery.webRTCStar.Enabled', false);
      set__default["default"](config, 'Swarm', {
        ...config.Swarm || {},
        DisableNatPortMap: true
      });
      return config;
    }
  },
  'default-networking': {
    description: 'Restores default network settings. Inverse profile of the `test` profile.',
    transform: config => {
      const defaultConfig = getDefaultConfig__default["default"]();
      set__default["default"](config, 'Addresses.API', defaultConfig.Addresses.API);
      set__default["default"](config, 'Addresses.Gateway', defaultConfig.Addresses.Gateway);
      set__default["default"](config, 'Addresses.Swarm', defaultConfig.Addresses.Swarm);
      set__default["default"](config, 'Addresses.Delegates', defaultConfig.Addresses.Delegates);
      set__default["default"](config, 'Bootstrap', defaultConfig.Bootstrap);
      set__default["default"](config, 'Discovery.MDNS.Enabled', defaultConfig.Discovery.MDNS.Enabled);
      set__default["default"](config, 'Discovery.webRTCStar.Enabled', defaultConfig.Discovery.webRTCStar.Enabled);
      set__default["default"](config, 'Swarm', {
        ...config.Swarm || {},
        DisableNatPortMap: false
      });
      return config;
    }
  },
  lowpower: {
    description: 'Reduces daemon overhead on the system. May affect node functionality,performance of content discovery and data fetching may be degraded. Recommended for low power systems.',
    transform: config => {
      const Swarm = config.Swarm || {};
      const ConnMgr = Swarm.ConnMgr || {};
      ConnMgr.LowWater = 20;
      ConnMgr.HighWater = 40;
      Swarm.ConnMgr = ConnMgr;
      config.Swarm = Swarm;
      return config;
    }
  },
  'default-power': {
    description: 'Inverse of "lowpower" profile.',
    transform: config => {
      const defaultConfig = getDefaultConfig__default["default"]();
      config.Swarm = defaultConfig.Swarm;
      return config;
    }
  }
};

exports.profiles = profiles;
