// ipfs-core-config/esm/src/libp2p.js
import TCP from 'libp2p-tcp';
import MulticastDNS from 'libp2p-mdns';
import WS from 'libp2p-websockets';
import KadDHT from 'libp2p-kad-dht';
import GossipSub from 'libp2p-gossipsub';
import Multiplex from 'libp2p-mplex';
import { NOISE } from '@chainsafe/libp2p-noise';
import {
  validator,
  selector
} from './utils/ipns.js';
import os from 'os';
export function libp2pConfig() {
  const options = {
    dialer: {
      maxParallelDials: 150,
      maxDialsPerPeer: 4,
      dialTimeout: 10000
    },
    modules: {
      transport: [
        TCP,
        WS
      ],
      streamMuxer: [Multiplex],
      connEncryption: [NOISE],
      peerDiscovery: [MulticastDNS],
      dht: KadDHT,
      pubsub: GossipSub
    },
    config: {
      peerDiscovery: {
        autoDial: true,
        [MulticastDNS.tag]: { enabled: true },
        bootstrap: { enabled: true }
      },
      dht: {
        kBucketSize: 20,
        enabled: true,
        clientMode: true,
        validators: { ipns: validator },
        selectors: { ipns: selector }
      },
      pubsub: {
        enabled: true,
        emitSelf: true
      },
      nat: {
        enabled: true,
        description: `ipfs@${ os.hostname() }`
      }
    },
    metrics: { enabled: true },
    peerStore: { persistence: true }
  };
  return options;
}
