// ipfs-core-config/esm/src/libp2p.browser.js
import WS from 'libp2p-websockets';
import WebRTCStar from 'libp2p-webrtc-star';
import Multiplex from 'libp2p-mplex';
import { NOISE } from '@chainsafe/libp2p-noise';
import KadDHT from 'libp2p-kad-dht';
import GossipSub from 'libp2p-gossipsub';
import {
  validator,
  selector
} from './utils/ipns.js';
export function libp2pConfig() {
  const options = {
    dialer: {
      maxParallelDials: 150,
      maxDialsPerPeer: 4,
      dialTimeout: 10000
    },
    modules: {
      transport: [
        WS,
        WebRTCStar
      ],
      streamMuxer: [Multiplex],
      connEncryption: [NOISE],
      peerDiscovery: [],
      dht: KadDHT,
      pubsub: GossipSub
    },
    config: {
      peerDiscovery: {
        autoDial: true,
        bootstrap: { enabled: true },
        webRTCStar: { enabled: true }
      },
      dht: {
        kBucketSize: 20,
        enabled: true,
        clientMode: true,
        validators: { ipns: validator },
        selectors: { ipns: selector }
      },
      pubsub: {
        enabled: true,
        emitSelf: true
      },
      nat: { enabled: false }
    },
    metrics: { enabled: true },
    peerStore: {
      persistence: true,
      threshold: 1
    }
  };
  return options;
}
