// ipfs-bitswap/cjs/src/utils/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var debug = require('debug');
var equals = require('uint8arrays/equals');
var entry = require('../message/entry.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var debug__default = /*#__PURE__*/_interopDefaultLegacy(debug);

const logger = (id, subsystem) => {
  const name = ['bitswap'];
  if (subsystem) {
    name.push(subsystem);
  }
  if (id) {
    name.push(`${ id.toB58String().slice(0, 8) }`);
  }
  return Object.assign(debug__default["default"](name.join(':')), { error: debug__default["default"](name.concat(['error']).join(':')) });
};
const includesWith = (pred, x, list) => {
  let idx = 0;
  const len = list.length;
  while (idx < len) {
    if (pred(x, list[idx])) {
      return true;
    }
    idx += 1;
  }
  return false;
};
const uniqWith = (pred, list) => {
  let idx = 0;
  const len = list.length;
  const result = [];
  let item;
  while (idx < len) {
    item = list[idx];
    if (!includesWith(pred, item, result)) {
      result[result.length] = item;
    }
    idx += 1;
  }
  return result;
};
const groupBy = (pred, list) => {
  return list.reduce((acc, v) => {
    const k = pred(v);
    if (acc[k]) {
      acc[k].push(v);
    } else {
      acc[k] = [v];
    }
    return acc;
  }, {});
};
const pullAllWith = (pred, list, values) => {
  return list.filter(i => {
    return !includesWith(pred, i, values);
  });
};
const sortBy = (fn, list) => {
  return Array.prototype.slice.call(list, 0).sort((a, b) => {
    const aa = fn(a);
    const bb = fn(b);
    return aa < bb ? -1 : aa > bb ? 1 : 0;
  });
};
const isMapEqual = (a, b) => {
  if (a.size !== b.size) {
    return false;
  }
  for (const [key, valueA] of a) {
    const valueB = b.get(key);
    if (valueB === undefined) {
      return false;
    }
    if (valueA instanceof Uint8Array && valueB instanceof Uint8Array && !equals.equals(valueA, valueB)) {
      return false;
    }
    if (valueA instanceof entry.BitswapMessageEntry && valueB instanceof entry.BitswapMessageEntry && !valueA.equals(valueB)) {
      return false;
    }
  }
  return true;
};

exports.groupBy = groupBy;
exports.includesWith = includesWith;
exports.isMapEqual = isMapEqual;
exports.logger = logger;
exports.pullAllWith = pullAllWith;
exports.sortBy = sortBy;
exports.uniqWith = uniqWith;
